### ISSUE TITLE
unsatisfiable constraints musl-1.1.15-r6 - edge APKINDEX incompatible with 3.5

### ISSUE THREAD

**Issue by IceDragon200** [bitwalker/alpine-erlang#8]
Due to the presence of the edge repo in the repositories, it now pulls in an incompatible musl lib.
I'm not sure if an update to Alpine 3.6 would solve this.

__Expected__

Packages to be installed successfully

__Actual__

```
+ apk update
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
fetch http://nl.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
v3.5.2-119-g586221122a [http://dl-cdn.alpinelinux.org/alpine/v3.5/main]
v3.5.2-110-gd9e62397fb [http://dl-cdn.alpinelinux.org/alpine/v3.5/community]
v3.6.0-1258-gf5668c29cd [http://nl.alpinelinux.org/alpine/edge/main]
OK: 13527 distinct packages available
+ apk --no-cache add build-base git make g++ wget curl inotify-tools nodejs
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
fetch http://nl.alpinelinux.org/alpine/edge/main/x86_64/APKINDEX.tar.gz
ERROR: unsatisfiable constraints:
  musl-1.1.15-r6:
    breaks: musl-dev-1.1.15-r7[musl=1.1.15-r7]
    satisfies: musl-utils-1.1.15-r6[musl=1.1.15-r6]
               musl-utils-1.1.15-r6[so:libc.musl-x86_64.so.1]
               pkgconf-1.0.2-r0[so:libc.musl-x86_64.so.1]
               wget-1.18-r2[so:libc.musl-x86_64.so.1]
               gcc-6.2.1-r1[so:libc.musl-x86_64.so.1]
               inotify-tools-3.14-r2[so:libc.musl-x86_64.so.1]
               binutils-libs-2.27-r1[so:libc.musl-x86_64.so.1]
               libcrypto1.0-1.0.2k-r0[so:libc.musl-x86_64.so.1]
               g++-6.2.1-r1[so:libc.musl-x86_64.so.1]
               readline-6.3.008-r4[so:libc.musl-x86_64.so.1]
               http-parser-2.7.1-r0[so:libc.musl-x86_64.so.1]
               mpfr3-3.1.5-r0[so:libc.musl-x86_64.so.1]
               libuv-1.9.1-r0[so:libc.musl-x86_64.so.1]
               ncurses-libs-6.0-r7[so:libc.musl-x86_64.so.1]
               libstdc++-6.2.1-r1[so:libc.musl-x86_64.so.1]
               nodejs-6.9.5-r0[so:libc.musl-x86_64.so.1]
               isl-0.17.1-r0[so:libc.musl-x86_64.so.1]
               libcurl-7.52.1-r3[so:libc.musl-x86_64.so.1]
               mpc1-1.0.3-r0[so:libc.musl-x86_64.so.1]
               libressl2.4-libcrypto-2.4.4-r0[so:libc.musl-x86_64.so.1]
               git-2.11.2-r0[so:libc.musl-x86_64.so.1]
               pcre-8.40-r2[so:libc.musl-x86_64.so.1]
               unixodbc-2.3.4-r0[so:libc.musl-x86_64.so.1]
               libressl2.4-libssl-2.4.4-r0[so:libc.musl-x86_64.so.1]
               zlib-1.2.8-r2[so:libc.musl-x86_64.so.1]
               libgomp-6.2.1-r1[so:libc.musl-x86_64.so.1]
               apk-tools-2.6.8-r2[so:libc.musl-x86_64.so.1]
               libssh2-1.7.0-r2[so:libc.musl-x86_64.so.1]
               libgcc-6.2.1-r1[so:libc.musl-x86_64.so.1]
               make-4.2.1-r0[so:libc.musl-x86_64.so.1]
               curl-7.52.1-r3[so:libc.musl-x86_64.so.1]
               alpine-baselayout-3.0.4-r0[so:libc.musl-x86_64.so.1]
               libatomic-6.2.1-r1[so:libc.musl-x86_64.so.1]
               ca-certificates-20161130-r1[so:libc.musl-x86_64.so.1]
               gmp-6.1.1-r0[so:libc.musl-x86_64.so.1]
               busybox-1.25.1-r0[so:libc.musl-x86_64.so.1]
               scanelf-1.1.6-r0[so:libc.musl-x86_64.so.1]
               expat-2.2.0-r0[so:libc.musl-x86_64.so.1]
               binutils-2.27-r1[so:libc.musl-x86_64.so.1]
               libssl1.0-1.0.2k-r0[so:libc.musl-x86_64.so.1]
```

**Comment by bitwalker** [comment]
I'm going to publish a new version of 19.3.4 with your PR merged, and then a new build with 20.0, then we can double check this PR. My understanding is that edge should only be getting used when explicitly labelled, or if a package is missing from the 3.4/3.6 repository, but if that's not the case, I'm open to suggestions on how to better use the two repositories in concert.

**Comment by idyll** [comment]
I have been fighting will my builds all day ever since this change. It's probably just me doing something wrong but I can't get brunch to work on the new 19.3.4 (3.6) image. It segfaults halfway through compiling things. There's no obvious culprit and it builds just fine on Ubuntu or macOS.

To get my image to build I have had to rollback to the old 6.1 image as 19 has the `musl` issue.

It would be super awesome if we could get the pre-3.6 image working again. Or any thoughts on what the difference between the alpine images would be. For example, I know that you need to install npm separate from nodejs now. I suspect there are other subtleties as well...

**Comment by bitwalker** [comment]
@idyll Are you using `alpine-erlang` directly or `alpine-elixir`? If the former, how are you installing and using node? I can try to do some testing and see what may be going on.

**Comment by idyll** [comment]
I am using alpine-erlang. And distillery (thanks!)

This is my current docker file:

```
FROM bitwalker/alpine-erlang:19.3.4

ENV HOME=/opt/app/ TERM=xterm

# Install Elixir and basic build dependencies
RUN \
    echo "@edge http://nl.alpinelinux.org/alpine/edge/community" >> /etc/apk/repositories && \
    apk update && \
    apk --no-cache --update add \
      git make g++ curl \
      nodejs-current nodejs-current-npm \
      python imagemagick \
      elixir@edge && \
    rm -rf /var/cache/apk/*

# Install Hex+Rebar
RUN mix local.hex --force && \
    mix local.rebar --force

WORKDIR /opt/app

ENV MIX_ENV=prod

# Cache node deps
COPY mix.exs mix.lock ./
COPY config/ config/
RUN mix do deps.get, deps.compile
COPY assets/ assets/
WORKDIR /opt/app/assets
RUN npm install
RUN node node_modules/brunch/bin/brunch build -d -p
WORKDIR /opt/app
COPY . .

RUN MIX_ENV=prod mix phx.digest
RUN mix release --env=prod --verbose
```

**Comment by idyll** [comment]
I have tried with `nodejs` and `nodejs-npm` as well.

**Comment by idyll** [comment]
Right now it's working if I update `musl` in the 19.3 image like this:

```
FROM bitwalker/alpine-erlang:19.3

ENV HOME=/opt/app/ TERM=xterm

# Install Elixir and basic build dependencies
RUN \
    echo "@edge http://nl.alpinelinux.org/alpine/edge/community" >> /etc/apk/repositories && \
    apk update && \
    apk --no-cache --update upgrade busybox musl && \
    apk --no-cache --update add \
      git make g++ curl \
      nodejs python elixir@edge && \
    rm -rf /var/cache/apk/*

# Install Hex+Rebar
RUN mix local.hex --force && \
    mix local.rebar --force

WORKDIR /opt/app

ENV MIX_ENV=prod

# Cache node deps
COPY mix.exs mix.lock ./
COPY config/ config/
RUN mix do deps.get, deps.compile
COPY assets/ assets/
WORKDIR /opt/app/assets
RUN npm install
RUN node node_modules/brunch/bin/brunch build -d -p
WORKDIR /opt/app
COPY . .

RUN MIX_ENV=prod mix phx.digest
RUN mix release --env=prod --verbose
```

**Comment by bitwalker** [comment]
Ah I see, the current Dockerfile doesn't have the `--update` flag for `upgrade`. I'll get that fixed.

**Comment by idyll** [comment]
I am still trying to determine the source of the segfault in the 3.6 image.... :(
If I track it down I will report back.

**Comment by bitwalker** [comment]
@idyll Shit, I just realized I didn't update the 19.3 tag to point to 19.3.4, I think that's your problem.

**Comment by bitwalker** [comment]
@idyll I've published 19.3 which points to 19.3.4

**Comment by idyll** [comment]
LOL. So @bitwalker it's pretty much the opposite issue I am having.

You merged in this PR which means since 19.3 -> 19.3.4, it NOW runs Alpine 3.6.
Brunch build segfaults on Alpine 3.6 (works just fine on 3.5, still trying to get to the bottom of it...)

I will just fork this until we get the 3.6 segfault sorted out so that there's still a version running alpine 3.5.

**Comment by bitwalker** [comment]
I've modified some things in how the image is created, so I think this issue should be resolved. I'm going to close it for now, but if it comes up again, open a new issue and we can address it fresh.
